File: pocs/rtk-poc/src/game.ts
export type Card = {
  id: number;
  symbol: string;
  matched: boolean;
};

const SYMBOLS = ["🍎", "🍌", "🍒", "🍇", "🍋", "🍓", "🍉", "🥝"];

export function buildDeck(): Card[] {
  const deck: Card[] = SYMBOLS.flatMap((symbol, i) => [
    { id: i * 2, symbol, matched: false },
    { id: i * 2 + 1, symbol, matched: false },
  ]);
  return shuffle(deck);
}

export function shuffle<T>(items: T[]): T[] {
  const copy = items.slice();
  for (let i = copy.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

export function allMatched(deck: Card[]): boolean {
  return deck.every((c) => c.matched);
}

File: pocs/rtk-poc/backend/src/game.rs
use rand::seq::SliceRandom;
use rand::thread_rng;
use serde::Serialize;

pub const SYMBOLS: [&str; 8] = ["🍎", "🍌", "🍒", "🍇", "🍋", "🍓", "🍉", "🥝"];

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct Card {
    pub id: u32,
    pub symbol: String,
}

pub fn build_deck() -> Vec<Card> {
    let mut deck: Vec<Card> = SYMBOLS
        .iter()
        .enumerate()
        .flat_map(|(i, s)| {
            vec![
                Card { id: (i as u32) * 2, symbol: s.to_string() },
                Card { id: (i as u32) * 2 + 1, symbol: s.to_string() },
            ]
        })
        .collect();
    deck.shuffle(&mut thread_rng());
    deck
}

pub fn pair_count(deck: &[Card]) -> usize {
    deck.len() / 2
}
